pkg/policy/l3.go

```diff
@@ -38,45 +39,70 @@ type L3PolicyMap struct {
 	IPv4Count   int // Count of IPv4 prefixes in 'Map'
 }
 
-// Insert places 'cidr' in to map 'm'. Returns `1` if `cidr` is added
-// to the map, `0` otherwise
-func (m *L3PolicyMap) Insert(cidr string) int {
-	_, ipnet, err := net.ParseCIDR(cidr)
-	if err != nil {
-		var mask net.IPMask
-		ip := net.ParseIP(cidr)
-		// Use default CIDR mask for the address if the bits in the address
-		// after the mask are all zeroes.
-		ip4 := ip.To4()
-		if ip4 == nil {
-			mask = net.CIDRMask(128, 128)
-		} else { // IPv4
-			ip = ip4
-			mask = ip.DefaultMask() // IP address class based mask (/8, /16, or /24)
-			if !ip.Equal(ip.Mask(mask)) {
-				// IPv4 with non-zeroes after the subnetwork, use full mask.
-				mask = net.CIDRMask(32, 32)
+// Insert places 'cidrs' in to map 'm'. Returns the number of new CIDRs added
+// to the map.
+func (m *L3PolicyMap) Insert(cidrs []string) int {
+
+	ipNets := []*net.IPNet{}
+	// Convert all provided CIDRs into net.IPNet
+	for _, cidr := range cidrs {
+		_, ipnet, err := net.ParseCIDR(cidr)
+		if err != nil {
+			var mask net.IPMask
+			ip := net.ParseIP(cidr)
+			// Use default CIDR mask for the address if the bits in the address
+			// after the mask are all zeroes.
+			ip4 := ip.To4()
+			if ip4 == nil {
+				mask = net.CIDRMask(128, 128)
+			} else { // IPv4
+				ip = ip4
+				mask = ip.DefaultMask() // IP address class based mask (/8, /16, or /24)
+				if !ip.Equal(ip.Mask(mask)) {
+					// IPv4 with non-zeroes after the subnetwork, use full mask.
+					mask = net.CIDRMask(32, 32)
+				}
 			}
+			ipnet = &net.IPNet{IP: ip, Mask: mask}
 		}
-		ipnet = &net.IPNet{IP: ip, Mask: mask}
+		ipNets = append(ipNets, ipnet)
+	}
+
+	// Grab all ipNets from existing map and coalesce them.
+	for _, v := range m.Map {
+		vv := v
+		ipNets = append(ipNets, &vv)
 	}
 
-	ones, _ := ipnet.Mask.Size()
+	count := 0
+	coalescedV4, coalescedV6 := ip.CoalesceCIDRs(ipNets)
 
-	key := ipnet.IP.String() + "/" + strconv.Itoa(ones)
-	if _, found := m.Map[key]; !found {
-		m.Map[key] = *ipnet
-		if ipnet.IP.To4() == nil {
-			m.IPv6Count++
-			m.IPv6Changed = true
-		} else {
+	for _, ipnet := range coalescedV4 {
+
+		ones, _ := ipnet.Mask.Size()
+
+		key := ipnet.IP.String() + "/" + strconv.Itoa(ones)
+		if _, found := m.Map[key]; !found {
+			m.Map[key] = *ipnet
 			m.IPv4Count++
 			m.IPv4Changed = true
+			count++
+		}
+	}
+
+	for _, ipnet := range coalescedV6 {
+		ones, _ := ipnet.Mask.Size()
+
+		key := ipnet.IP.String() + "/" + strconv.Itoa(ones)
+		if _, found := m.Map[key]; !found {
+			m.Map[key] = *ipnet
+			m.IPv6Count++
```

Comment:
I'm intrigued by the semantics of these `IPv4Count` and `IPv6Count` fields. The resulting values don't make sense to me, as they are never set to zero anywhere. Shouldn't they be set to zero before the loops over `coalescedV4` and `coalescedV6`?
I think you should just do:
```
m.IPv4Count = len(coalescedV4)
m.IPv6Count = len(coalescedV6)
```